neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsListIterator.java

```diff
@@ -0,0 +1,594 @@
+     *
+     * @see #feature()
+     */
+    protected EStructuralFeature currentFeature;
+
+    /**
+     * {@code true} if the prepared result is a feature map.
+     */
+    protected boolean isFeatureMap;
+
+    /**
+     * The list iterator on values; only with multi-valued features and feature maps.
+     */
+    protected ListIterator<E> values;
+
+    /**
+     * The feature filter to determine if a feature should be ignored from the result.
+     */
+    @Nullable
+    protected EContentsEList.FeatureFilter filter;
+
+    /**
+     * Constructs a new {@code ContentsListIterator} for the given {@code owner}.
+     *
+     * @param owner    the owner of this iterator
+     * @param features the containment features that are handled by this iterator
+     */
+    public ContentsListIterator(PersistentEObject owner, EStructuralFeature[] features) {
+        this.owner = owner;
+        this.features = Arrays.asList(features).listIterator();
+    }
+
+    /**
+     * Returns {@code true} if proxied {@link EObject objects} has to be resolved.
+     *
+     * @return {@code true} if proxied objects has to be resolved
+     */
+    protected boolean resolve() {
+        return false;
+    }
+
+    /**
+     * Returns {@code true} if the {@code feature} must be included in the result, {@code false} if it should be ignored.
+     *
+     * @param feature the feature to test
+     *
+     * @return {@code true} if the {@code feature} must be included in the result, {@code false} if it should be ignored
+     *
+     * @see EContentsEList.FeatureFilter#isIncluded(EStructuralFeature)
+     */
+    protected boolean isIncluded(EStructuralFeature feature) {
+        return isNull(filter) || filter.isIncluded(feature);
+    }
+
+    /**
+     * Returns {@code true} if the feature of the {@code entry} must be included in the result, {@code false} if it
+     * should be ignored.
+     *
+     * @param entry the entry to test
+     *
+     * @return {@code true} if the feature of the {@code entry} must be included in the result, {@code false} if it
+     * should be ignored
+     *
+     * @see #isIncluded(EStructuralFeature)
+     */
+    protected boolean isIncludedEntry(FeatureMap.Entry entry) {
+        final EStructuralFeature f = entry.getEStructuralFeature();
+
+        return nonNull(entry.getValue())
+                && isIncluded(f)
+                && EObjects.isReference(f)
+                && EObjects.asReference(f).isContainment();
+    }
+
+    @Override
+    public void filter(EContentsEList.FeatureFilter newFilter) {
+        checkState(cursor == 0 && isNull(filter), "Iterator already in use or already filtered");
+        filter = newFilter;
+    }
+
+    @Override
+    public EStructuralFeature feature() {
+        return currentFeature;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return hasMoreElements(IterationHelpers.ASCENDING);
+    }
+
+    @Nonnull
+    @Override
+    public E next() {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Add a "NoSuchElementException" for iteration beyond the end of the collection. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2272)